File: src/content/2.4/code/scala/snippet15.scala
implicit val streamRepresentable = new Representable[Stream] {
  type Rep = Int

  def tabulate[X](f: Rep => X): Stream[X] = 
    Stream[X](() => f(0), () => tabulate(f compose (_ + 1)))

  def index[X]: Stream[X] => Rep => X = {
    case Stream(b, bs) => 
      n =>
        if (n == 0) b()
        else index(bs())(n - 1)
  }
}

File: src/content/2.4/code/ocaml/snippet15.ml
module StreamRepresentable : Representable = struct
  type rep = int
  type 'x t = 'x stream

  let rec tabulate f = Cons (f 0, lazy (tabulate (compose f succ)))

  let rec index (Cons (b, bs)) n =
    if n = 0 then b else index (Lazy.force bs) (n - 1)
  ;;
end
